java-checks/src/main/java/org/sonar/java/checks/MembersDifferOnlyByCapitalizationCheck.java

```diff
@@ -96,7 +96,32 @@ private static boolean isOverriding(Symbol symbol) {
 
   private static boolean isInvalidMember(Symbol currentMember, Symbol knownMember) {
     if (!isOverriding(currentMember)) {
-      return differentTypes(currentMember, knownMember) ? sameVisibilityNotPrivate(currentMember, knownMember) : !sameName(currentMember, knownMember);
+      return differentTypes(currentMember, knownMember) ? invalidMethodAndVariable(currentMember, knownMember) : !sameName(currentMember, knownMember);
+    }
+    return false;
+  }
+
+  private static boolean invalidMethodAndVariable(Symbol currentMember, Symbol knownMember) {
+    Symbol methodSymbol = currentMember.isMethodSymbol() ? currentMember : knownMember;
+    Symbol variableSymbol = methodSymbol == currentMember ? knownMember : currentMember;
+    return sameVisibilityNotPrivate(currentMember, knownMember)
+      && !methodReturningVariableWithSameName(methodSymbol, variableSymbol)
+      && !isBuilderPattern(methodSymbol, variableSymbol);
+  }
+
+  private static boolean isBuilderPattern(Symbol methodSymbol, Symbol variableSymbol) {
+    return methodSymbol.owner().name().endsWith("Builder") && sameName(methodSymbol, variableSymbol);
+  }
+
+  private static boolean methodReturningVariableWithSameName(Symbol methodSymbol, Symbol variableSymbol) {
+    if (!sameName(variableSymbol, methodSymbol)) {
+      return false;
+    }
+    Tree declaration = methodSymbol.declaration();
+    if (declaration != null) {
+      ReturnVisitor returnVisitor = new ReturnVisitor(variableSymbol);
+      declaration.accept(returnVisitor);
+      return returnVisitor.singleReturnWithVariableSymbol();
```

Comment:
Good point! I excluded anonymous classes and lambdas, and added corresponding tests.